// fumadocs-core/dist/highlight/core/index.js
import { Fragment } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { toJsxRuntime } from "hast-util-to-jsx-runtime";
//#region src/highlight/core/index.ts
async function highlightHast(code, options) {
	const { fallbackLanguage = "text", config, ...resolved } = options;
	Object.assign(resolved, config.resolveThemes(resolved));
	let themesToLoad;
	if ("theme" in resolved) themesToLoad = [resolved.theme];
	else if ("themes" in resolved) themesToLoad = Object.values(resolved.themes).filter((v) => v !== void 0);
	else throw new Error("impossible");
	const { isSpecialLang } = await import("shiki/core");
	const highlighter = await config.createHighlighter();
	if (!isSpecialLang(resolved.lang) && !(resolved.lang in highlighter.getBundledLanguages()) && !highlighter.getLoadedLanguages().includes(resolved.lang)) resolved.lang = fallbackLanguage;
	await Promise.all([loadMissingTheme(highlighter, ...themesToLoad), loadMissingLanguage(highlighter, resolved.lang)]);
	return highlighter.codeToHast(code, resolved);
}
async function loadMissingTheme(highlighter, ...themes) {
	const { isSpecialTheme } = await import("shiki/core");
	const missingThemes = themes.filter((theme) => {
		if (isSpecialTheme(theme)) return false;
		try {
			highlighter.getTheme(theme);
			return false;
		} catch {
			return true;
		}
	});
	if (missingThemes.length > 0) await highlighter.loadTheme(...missingThemes);
}
async function loadMissingLanguage(highlighter, ...langs) {
	const { isSpecialLang } = await import("shiki/core");
	const missingLangs = langs.filter((lang) => {
		if (isSpecialLang(lang)) return false;
		try {
			highlighter.getLanguage(lang);
			return false;
		} catch {
			return true;
		}
	});
	if (missingLangs.length > 0) await highlighter.loadLanguage(...missingLangs);
}
/**
* Get Shiki highlighter instance of Fumadocs (mostly for internal use, you should use Shiki directly over this).
*
* @param engineType - Shiki Regex engine to use.
* @param options - Shiki options.
*/
async function getHighlighter(config, options) {
	const highlighter = await config.createHighlighter();
	await Promise.all([options?.langs && loadMissingLanguage(highlighter, ...options.langs), options?.themes && loadMissingTheme(highlighter, ...options.themes)]);
	return highlighter;
}
async function highlight(code, options) {
	return toJsxRuntime(await highlightHast(code, options), {
		jsx,
		jsxs,
		development: false,
		Fragment,
		components: options.components
	});
}
//#endregion
export { getHighlighter, highlight, highlightHast };
