// fumadocs-core/dist/highlight/index.js
import { getHighlighter as getHighlighter$1, highlight as highlight$1, highlightHast as highlightHast$1 } from "./core/index.js";
import { defineShikiConfig } from "./config.js";
//#region src/highlight/index.ts
async function highlightHast(code, options) {
	const engine = options.engine ?? "js";
	return highlightHast$1(code, {
		...options,
		config: engine === "js" ? configDefault : configWASM
	});
}
/**
* Get Shiki highlighter instance of Fumadocs (mostly for internal use, you should use Shiki directly over this).
*
* @param engineType - Shiki Regex engine to use.
* @param options - Shiki options.
*/
async function getHighlighter(engineType, options) {
	return getHighlighter$1(engineType === "js" ? configDefault : configWASM, options);
}
async function highlight(code, options) {
	const engine = options.engine ?? "js";
	return highlight$1(code, {
		...options,
		config: engine === "js" ? configDefault : configWASM
	});
}
const defaultThemes = { themes: {
	light: "github-light",
	dark: "github-dark"
} };
const configDefault = defineShikiConfig({
	defaultThemes,
	async createHighlighter() {
		const { createHighlighter } = await import("shiki");
		const { createJavaScriptRegexEngine } = await import("shiki/engine/javascript");
		return createHighlighter({
			langs: [],
			themes: [],
			engine: createJavaScriptRegexEngine()
		});
	}
});
/** config using the WASM powered Regex engine */
const configWASM = defineShikiConfig({
	defaultThemes,
	async createHighlighter() {
		const { createHighlighter, createOnigurumaEngine } = await import("shiki");
		return createHighlighter({
			langs: [],
			themes: [],
			engine: createOnigurumaEngine(import("shiki/wasm"))
		});
	}
});
//#endregion
export { configDefault, configWASM, getHighlighter, highlight, highlightHast };
